# **Предобработка данных**

Работа с пропусками. Определение аномальных значений. Преобразование типов данных. Основные методы поиска дубликатов. Работа с несовершенными реальными наборами данных. Проект. Проверить гипотезу о влиянии email-рассылки на динамику повторных покупок.

# **Описание проекта**
Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.
Результаты исследования будут учтены при построении модели кредитного скоринга — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.
# **Описание данных**

    children — количество детей в семье
    days_employed — общий трудовой стаж в днях
    dob_years — возраст клиента в годах
    education — уровень образования клиента
    education_id — идентификатор уровня образования
    family_status — семейное положение
    family_status_id — идентификатор семейного положения
    gender — пол клиента
    income_type — тип занятости
    debt — имел ли задолженность по возврату кредитов
    total_income — ежемесячный доход
    purpose — цель получения кредита

# Исследование надежности заемщиков


## Откройте таблицу и изучите общую информацию о данных

**Импортируем библиотеку pandas. Считаем данные из csv-файла в датафрейм и сохраним в переменную `data`.** 


In [ ]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Выведем первые 20 строчек датафрейма `data` на экран.**

In [ ]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Выведем основную информацию о датафрейме с помощью метода `info()`.**

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Выведем количество пропущенных значений для каждого столбца. Используем комбинацию двух методов.**

In [ ]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

In [ ]:
#создание таблицы, в столбцах days_employed, total_income которой отсутствуют значения. Для последних вопросов.
data_nan = data.loc[(data['total_income'].isna()) & (data['days_employed'].isna())]
data_nan

children  days_employed  dob_years education  education_id  \
12            0            NaN         65   среднее             1   
26            0            NaN         41   среднее             1   
29            0            NaN         63   среднее             1   
41            0            NaN         50   среднее             1   
55            0            NaN         54   среднее             1   
...         ...            ...        ...       ...           ...   
21489         2            NaN         47   Среднее             1   
21495         1            NaN         50   среднее             1   
21497         0            NaN         48    ВЫСШЕЕ             0   
21502         1            NaN         42   среднее             1   
21510         2            NaN         28   среднее             1   

               family_status  family_status_id gender  income_type  debt  \
12          гражданский брак                 1      M    пенсионер     0   
26           женат / замужем                 0      M  госслужащий     0   
29     Не женат / не замужем                 4      F    пенсионер     0   
41           женат / замужем                 0      F  госслужащий     0   
55          гражданский брак                 1      F    пенсионер     1   
...                      ...               ...    ...          ...   ...   
21489        женат / замужем                 0      M    компаньон     0   
21495       гражданский брак                 1      F    сотрудник     0   
21497        женат / замужем                 0      F    компаньон     0   
21502        женат / замужем                 0      F    сотрудник     0   
21510        женат / замужем                 0      F    сотрудник     0   

       total_income                           purpose  
12              NaN                   сыграть свадьбу  
26              NaN                       образование  
29              NaN  строительство жилой недвижимости  
41              NaN  сделка с подержанным автомобилем  
55              NaN                   сыграть свадьбу  
...             ...                               ...  
21489           NaN              сделка с автомобилем  
21495           NaN                           свадьба  
21497           NaN        строительство недвижимости  
21502           NaN  строительство жилой недвижимости  
21510           NaN           приобретение автомобиля  

[2174 rows x 12 columns]

**В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце обработаем на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [ ]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработаем значения в этом столбце: заменим все отрицательные значения положительными с помощью метода `abs()`.**

In [ ]:
data['days_employed'] = data['days_employed'].abs()

**Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [ ]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставим их как есть. Тем более этот столбец не понадобится нам для исследования.

**Выведите перечень уникальных значений столбца `children`.**

In [ ]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**В столбце `children` есть два аномальных значения. Удалим строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [ ]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Ещё раз выведем перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [ ]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Заполним пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [ ]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Убедимся, что все пропуски заполнены. Проверим себя и ещё раз выведем количество пропущенных значений для каждого столбца с помощью двух методов.**

In [ ]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Заменим вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [ ]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Обработаем неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведем их к нижнему регистру. Проверим остальные столбцы.**

In [ ]:
data['education'] = data['education'].str.lower()

**Выведем на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалим их.**

In [ ]:
data.duplicated().sum()

71

In [ ]:
data = data.drop_duplicates()

### Категоризация данных

**На основании диапазонов, указанных ниже, создадим в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используем собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [ ]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [ ]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Выведем на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [ ]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Создадим функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используем собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучим данные в столбце `purpose` и определим, какие подстроки помогут нам правильно определить категорию.**

In [ ]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [ ]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуем данные и ответим на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [ ]:
#выведем сводную таблицу по количеству детей
#посчитаем число заемщиков, сумму неплатильщиков и их долю от общего числа берущих кредит
children_pivot = data.pivot_table(index=['children'],  values = 'debt', aggfunc = ['count', 'sum', 'mean'])

children_pivot

count   sum      mean
           debt  debt      debt
children                       
0         14091  1063  0.075438
1          4808   444  0.092346
2          2052   194  0.094542
3           330    27  0.081818
4            41     4  0.097561
5             9     0  0.000000

**Вывод:** В сводной таблице мы  видим, что больше всего людей, которые берут кредит, **не имеют детей**. Далее с увеличением количества детей уменьшается и число заемщиков. 
Из них во втором столбце **sum debt** наблюдаем самое большое число тех, кто не возвращает вовремя кредит относится к людям **не имеющих детей**. Остальные зачения должников также уменьшаются с увеличением числа детей. 
Третья колонка **mean debt** - рассказывает нам о том, что отношение числа заемщиков, не возвращающих вовремя кредит, к общему числу людей, взявших кредиты имеет, напротив, меньшее значение у категории людей **не имеющих детей**. У групп с **одним и более детьми** доля неплатильщиков больше, чем у заемщиков **без детей**, и имеет приблизительно одинаковое значение. Единственно у людей, имеющих **5 детей**, количество взявших кредит мало, всего 9, но и должников нет.

Отсюда можно сделать вывод, что  **зависимость между количеством детей и возвратом кредита в срок есть, но по задолженности число невовремя оплачивающих кредит меньше всего у  заемщиков, не имеющих детей**

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [ ]:
#выведем сводную таблицу по семейному положению
#посчитаем число заемщиков, сумму неплатильщиков и их долю от общего числа берущих кредит
family_status_pivot = data.pivot_table(index=['family_status'],  values = 'debt', aggfunc = ['count','sum', 'mean'])
family_status_pivot 

count  sum      mean
                        debt debt      debt
family_status                              
Не женат / не замужем   2796  273  0.097639
в разводе               1189   84  0.070648
вдовец / вдова           951   63  0.066246
гражданский брак        4134  385  0.093130
женат / замужем        12261  927  0.075606

**Вывод:** В данной сводной таблице больше всего заемщиков с семейны положением **замужем\женат**, как и колличество неплатильщиков по второй колонке **sum debt**. 
Исходя из третей колонки **mean debt**, доля людей, несовершающих вовремя платежи по кредиту, меньше всего среди тех, кто имеет статус **вдовец\вдова**, далее **в разводе**, и чуть больше среди **женат\замужем**. 
Остальные:**гражданский брак** и **не женат\не замужем** имеют наибольшее значение среди других  неплатильщиков.
Отсюда можем сделать выввод, что **связь между семейным положением и возвратом кредита в срок имеется.**

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [ ]:
#выведем сводную таблицу по категориям дохода
#посчитаем число заемщиков, сумму неплатильщиков и их долю от общего числа берущих кредит
total_income_category_pivot = data.pivot_table(index=['total_income_category'], \
                                               values = 'debt', aggfunc = ['count','sum', 'mean']) 
total_income_category_pivot


count   sum      mean
                        debt  debt      debt
total_income_category                       
A                         25     2  0.080000
B                       5014   354  0.070602
C                      15921  1353  0.084982
D                        349    21  0.060172
E                         22     2  0.090909

**Вывод:** Таблица составлена по категорям в соответствии с заработком. Максимальное число заемщиков в секторе **C** с доходом от 50 001 до 200 000, но и число неплатильщиков и их доля среди общего количества велико. 
Самое маленькое значение доли не возвращающих вовремя кредит у людей с доходом **D** от 30 001 до 50 000, далее чуть больше значение доли неплатильщиков у категории **В** с доходом от 200 001 до 1000 000. 
Думаю, это наиболее интересная для банка группа людей **В**,тк.  Количество людей, берущих кредит велико - на втором месте по значению общего числа, но и маленькая доля неплатильщиков.
Категория, имеющая самое большое количество, задерживающих платеж, это с доходом от 0 до 30 000 **Е**, но и количество берущих кредит мало, всго 22 человека. 

Отсюда сделаем вывод, что  **есть  зависимость между уровнем дохода и возвратом кредита в срок**, наиболее привлекательны категории с доходом **D** и **B**.

p.s. Однако 2174 пропуска в данных столбца "total_income" ставят под сомнение полученный нами результат. Сводная таблица (или группировка) с данными о зависимости между уровнем дохода и возвратом кредита в срок возможно могла бы выглядеть иначе, если бы не пропущенные данные.

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [ ]:
#выведем сводную таблицу по целям кредита
#посчитаем число заемщиков, сумму неплатильщиков и их долю от общего числа берущих кредит
purpose_category_pivot = data.pivot_table(index=['purpose_category'], values = 'debt', aggfunc = ['count','sum','mean'])
purpose_category_pivot

count  sum      mean
                           debt debt      debt
purpose_category                              
операции с автомобилем     4279  400  0.093480
операции с недвижимостью  10751  780  0.072551
получение образования      3988  369  0.092528
проведение свадьбы         2313  183  0.079118

**Вывод:** Перед нами сводная таблица с количеством людей, имеющие разные цели кредита,  и доли неплатильщиков. Видно, что наибольшее число людей берут кредит на **операции с недвижимостью**, среди них имеются и неплатильщики, доля которых наименьшая, по сравнению с остальными. 
Далее по общему числу взявших кредит идут цели **операции с автомобилем и получение образования**, среди которых доля невозвращающих  вовремя заем самая наибольшая.
На **проведение свадьбы** берет самое маленькое количество людей, доля вносивших платеж невовремя  имеет среднее значение среди всех остальных. 
Отсюда можем сделать вывод, что **имеется зависимость разных целей кредита и возвратом его в срок.**

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

In [ ]:
#выше была создана таблица, в столбцах days_employed, total_income которой отсутствуют значения. 
data_nan

children  days_employed  dob_years education  education_id  \
12            0            NaN         65   среднее             1   
26            0            NaN         41   среднее             1   
29            0            NaN         63   среднее             1   
41            0            NaN         50   среднее             1   
55            0            NaN         54   среднее             1   
...         ...            ...        ...       ...           ...   
21489         2            NaN         47   Среднее             1   
21495         1            NaN         50   среднее             1   
21497         0            NaN         48    ВЫСШЕЕ             0   
21502         1            NaN         42   среднее             1   
21510         2            NaN         28   среднее             1   

               family_status  family_status_id gender  income_type  debt  \
12          гражданский брак                 1      M    пенсионер     0   
26           женат / замужем                 0      M  госслужащий     0   
29     Не женат / не замужем                 4      F    пенсионер     0   
41           женат / замужем                 0      F  госслужащий     0   
55          гражданский брак                 1      F    пенсионер     1   
...                      ...               ...    ...          ...   ...   
21489        женат / замужем                 0      M    компаньон     0   
21495       гражданский брак                 1      F    сотрудник     0   
21497        женат / замужем                 0      F    компаньон     0   
21502        женат / замужем                 0      F    сотрудник     0   
21510        женат / замужем                 0      F    сотрудник     0   

       total_income                           purpose  
12              NaN                   сыграть свадьбу  
26              NaN                       образование  
29              NaN  строительство жилой недвижимости  
41              NaN  сделка с подержанным автомобилем  
55              NaN                   сыграть свадьбу  
...             ...                               ...  
21489           NaN              сделка с автомобилем  
21495           NaN                           свадьба  
21497           NaN        строительство недвижимости  
21502           NaN  строительство жилой недвижимости  
21510           NaN           приобретение автомобиля  

[2174 rows x 12 columns]

In [ ]:
#посчитаем частоту уникальных значений в столбце 'income_type'
data_nan['income_type'].value_counts()

сотрудник          1105
компаньон           508
пенсионер           413
госслужащий         147
предприниматель       1
Name: income_type, dtype: int64

In [ ]:
#посчитаем число выплачивающих вовремя и должников в столбце 'debt'
data_nan['debt'].value_counts()

0    2004
1     170
Name: debt, dtype: int64

In [ ]:
#приведем цели кредита к единым значениям
data_nan['purpose_category'] = data_nan['purpose'].apply(categorize_purpose)
data_nan['purpose_category'].value_counts()

операции с недвижимостью    1082
получение образования        425
операции с автомобилем       418
проведение свадьбы           249
Name: purpose_category, dtype: int64

Была создана таблица, в столбцах days_employed, total_income которой отсутствуют значения. Посчитала частоту уникальных значений в столбце 'income_type'. Видно, что встречаются разные типы занятости. Из них Сотрудник встречается чаще всего.

Посчитаем число выплачивающих вовремя и должников в столбце 'debt'. Большое всего выплачивающих вовремя, но и должники имеются.

Цели кредита также разные у клиентов с пропусками, максимальное количество у берущих кредит на операции с недвижимостью. При осмотре таблице также видны разные значения и в других столбцах.

Возможно, пропуски имеют клиенты, которые имеют карточки, обслуживающиеся именно этим банком, например, зарплатные, либо берущие кредит - это работники или партнеры этого банка, либо они предоставляют некий залог в качестве гарантий выплаты кредита банку, информации о которых нам неизвестна. 

В наших данных пропуски в столбцах "трудовой стаж" и "ежемесячный доход" абсолютно идентичны. Количество строк с отсутствующими данными в обоих указанных столбцах совпадает и равно 2174. Маловероятно, что специалисты по выдаче кредитов, заполнив все без исключения ячейки других столбцов по каждому кредитополучателю, оставляли незаполненными именно два показателя, и сделали это ровно 2174 раза в каждом столбце. 

Даже не проводя проверку, во всех ли случаях незаполненные ячейки находятся в одной и то же строке с высокой долей вероятности можно утверждать, что произошла техническая ошибка при формировании таблицы с исходными данными (либо некий сбой во время выгрузки данных на анализ, либо при объединении данных в единый файл, который также должен был передаваться на анализ, либо эти данные не собирались).

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* Медианое значение делит выборку значений пополам. На среднее арифметическое же очень сильно влияют очень большие или очень маленькие значения в наборе данных, выбросы. Соответственно, по итогу будет либо завышенное значение, либо заниженное. Медиана лишена влияния таких выбросов.

### Шаг 4: общий вывод.

 Из полученных данных можно сделать вывод, что **есть зависимость между количеством детей и выплат кредита в срок.** Наибольшая категория людей берущая кредит не имеет детей, и среди нее же наименьшая доля заемщиков, вовремя не выплачивающих платежи. Далее с увеличением количества детей уменьшается число берущих кредиты , а вот доля неплатильщиков приблизительно одинакова высока среди тех, кто имеет детей. 

Также  **зависимость между семейным положением и возвратом кредита в срок есть.** Во-первых, и число берущих кредит больше у людей с семейным положением **женат\замужем**, а доля задолжников одна из самых маленьких. Те, кто в **гражданском браке или не замужем\не женаты**, иемют наибольшее число должников. **В разводе и вдовы\вдовцы** берут кредит меньше раз, но и должников среди них самое маленькое количество.

Рассмотрены также заемщики по уровню дохода. Отмеченно, что наибольшее число берущих кредит среди категории людей **С** с уровнем дохода от 50 001 до 200 000, но и доля должников высока. Вторая категоря по общему  количеству заемщиков это **В** с уровнем дохода от 200 001 до 1000 000, среди них доля неплатильщиков одна из наименьших. С самой маленькой долей должников категория **D** с доходом от 30 001 до 50 000, но и общее количество берущих кредит мало. Категория **А** и **Е** имеет наименьшее количество людей с кредитом и высокие доли неплатильщиков. Соответственно, **есть зависимость между уровнем дохода и возвратом кредита в срок.** 

**Разные цели кредита также влияют на его возврат в срок**. Так наименьшая доля неплательщиков наблюдается среди людей взявших кредит **на операции с недвижимостью**, также и общее число берущих для этой цели самое большое. 

Чуть больше доля среди неплатильщиков с целью кредита **проведение свадьбы**, но общая доля заемщиков самая маленькая. Остальные категории целей кредита: **получение образования и операции с автомобилем** имеют приблизительно одинаковое количество заемщиков и одинаково высокую долю	неплатильщиков.